util/sock_util.c

```diff
@@ -276,3 +278,76 @@ sock_addr_prettyprint(const struct sock_addr * sa)
 		return (strdup("Unknown address"));
 	}
 }
+
+/**
+ * sock_addr_ensure_port(addr):
+ * Allocate a new string to serve as the address for sock_resolve().
+ * If ${addr} contains a port number or is the address of a Unix domain
+ * socket, duplicate that string; if not, add a port number of ":0".
+ */
+char *
+sock_addr_ensure_port(const char * addr)
+{
+	char * bind_addr;
+	char * cl;
+	char * cr;
+
+	/* Sanity check. */
+	assert(addr != NULL);
+
+	/* Where are the left-most and right-most colons in $addr? */
+	cl = strchr(addr, ':');
+	cr = strrchr(addr, ':');
+
+	/* If the right-most colon is the left-most character, bail. */
+	if (cr == addr) {
+		warn0("Invalid address: %s", addr);
+		goto err0;
+	}
+
+	/* Figure out what type of address $addr is. */
+	if (addr[0] == '/') {
+		/* It's a Unix domain socket and doesn't need a port number. */
+		bind_addr = strdup(addr);
+	} else if (addr[0] != '[') {
+		/* It's a hostname... */
+		if (cl == NULL) {
+			/* ... without any port number, so we add ":0". */
+			if (asprintf(&bind_addr, "%s:0", addr) == -1) {
+				warnp("asprintf");
+				goto err0;
+			}
+		} else {
+			/* ... which already has a port number. */
+			bind_addr = strdup(addr);
+		}
+	} else {
+		/* addr[0] == '[', so it's an address... */
+		if (cr == NULL) {
+			/* ... ipv4, without a port number, so we add ":0". */
+			if (asprintf(&bind_addr, "%s:0", addr) == -1) {
+				warnp("asprintf");
+				goto err0;
+			}
+		} else if (cl == cr) {
```

Comment:
You know this could also be `cr[-1] == ']'`.

And if you collapse those two, then you can collapse the two "without a port number" cases too:

```
if ((cr == NULL) || (cr[-1] != ']')) {
    /* IPv[46] address without a port number; add ":0". */
    ....
} else {
    /* IPv[46] address with a port number. */
    ...
}
```

Reply:
Ok, changed.  I used different comments (matching ones in the "hostname" section), but the code is the same as you suggested.

Shall I rebase this PR?